rfcs/0000-structured-logging.md

```diff
@@ -0,0 +1,2189 @@
+    ^^^^
+    structured
+);
+```
+
+### What can be logged?
+
+Anything that's `std::fmt::Debug + serde::Serialize` can be logged as a structured value. Using default crate features though, only a fixed set of types from the standard library are supported:
+
+- Standard formats: `Arguments`
+- Primitives: `bool`, `char`
+- Unsigned integers: `u8`, `u16`, `u32`, `u64`, `u128`
+- Signed integers: `i8`, `i16`, `i32`, `i64`, `i128`
+- Strings: `&str`, `String`
+- Bytes: `&[u8]`, `Vec<u8>`
+- Paths: `Path`, `PathBuf`
+- Special types: `Option<T>` and `()`
+
+This keeps the `log` crate lightweight by default. In the example from before, `correlation_id` and `user` can be used as structured values if they're in that set of types:
+
+```rust
+let user = "a user id";
+let correlation_id = "some correlation id";
+
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+What if the `correlation_id` is a `uuid::Uuid` instead of a string? What if the `user` is some other datastructure containing an id along with some other metadata? We still want to be able to log these values. The set of loggable values above can be expanded to also include any other type that implements both `std::fmt::Debug` and `serde::Serialize` using a crate feature:
+
+```toml
+[dependencies.log]
+features = ["kv_serde"]
+
+[dependencies.uuid]
+features = ["serde"]
+```
+
+```rust
+#[derive(Debug, Serialize)]
+struct User {
+    id: Uuid,
+    ..
+}
+
+let user = User { id, .. };
+let correlation_id = Uuid::new_v4();
+
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+    correlation = correlation_id,
+    user
+);
+```
+
+So the effective trait bounds for structured values are `Debug + Serialize`:
+
+```
+info!(
+    "This is the rendered {message}. It is not structured",
+    message = "message";
+              ^^^^^^^^^
+              Display
+
+    correlation = correlation_id,
+                  ^^^^^^^^^^^^^^
+                  Debug + Serialize
+
+    user
+    ^^^^
+    Debug + Serialize
+);
+```
+
+If you come across a data type in the Rust ecosystem that you can't log, then try looking for a `serde` feature on the crate that data type comes from. If there isn't one already then adding it will be useful not just for you, but for anyone that might want to serialize those types for other reasons.
+
+## Supporting key-value pairs in `Log` implementations
+
+Capturing structured logs is only half the story. Implementors of the `Log` trait also need to be able to work with any key-value pairs associated with a log record. Take the terminal log format from before:
+
+```
+[INF 2018-09-27T09:32:03Z] Operation completed successfully in 18ms
+module: "basic"
+service: "database"
+correlation: 123
+took: 18
+```
+
+Each key-value pair, shown as `$key: $value`, can be formatted using the `std::fmt` machinery:
+
+```rust
+use log::kv::Source;
```

Comment:
You're using a trait here that not has been yet in this RFC. Maybe put the definition of it before using it, or at least refer to the trait definition. Reading this top-to-bottom made it somewhat surprising to find this here.